### ISSUE TITLE
Norwegian has two different locales,  `nb_NO` and `nn_NO`

### ISSUE THREAD

**Issue by o-l-a-v** [geany/geany#3638]
Norwegian has two locales:

* `nb_NO`: Norwegian Bokmål, most used.
* `nn_NO`: Norwegian Nynorsk, least used.

Currently Geany has `nn` for norwegian nynorsk (`nn_NO`). We also have `nb`, norwegian bokmål (`nb_NO`).

Please add `nb_NO`, rename `nn` to `nn_NO`, and don't choose `nn_NO` in Geany if OS system language is set to `nb_NO`.

**Comment by elextr** [comment]
Probably would be good if "somebody" provided a preliminary `nb.po` file rather than rushing the whole translation at string freeze.

**Comment by b4n** [comment]
I'm not sure I entirely get this.  We have `nn` translation indeed.  We don't have `nb`.
I am not sure how the locale selection is made for Norwegian (I guess when the configured language is either `no_NO` or `nb_NO`) as we rely on the system/on the gettext library to perform this step; but from your report I can *guess* that it falls back on `nn` if there's no `nb`.

The easy solution would be to just submit an `nb` translation I guess?

> Please add `nb_NO`

This would require somebody contributing such a translation. And I would think it should be `nb` (without region) preferably (see below).

> rename `nn` to `nn_NO`

Why?  Is the translation we have really specific to the Norwegian variant of Nynorsk?

From research on my system there is *no* other common region for either `nb` or `nn`, and the norm seems to be not to add it:

| Language | Number of translations on my system |
|--------|--------|
| nb | 356 |
| nn | 292 |
| nb_NO | 10 |
| nn_NO | 1 |
| no | 4 | 
| no_NO | 0 | 

Additionally, there's only 242 software that have both an `nb` and `nn` translations, which means there's 50 that don't provide an `nb` one.  This doesn't mean anything relevant though, apart that most software that provide `nn` also provide `nb`.

> and don't choose `nn_NO` in Geany if OS system language is set to `nb_NO`.

Again, I think that would likely solve itself if we had a `nb` translation.

Also, would it really be preferable to expose an untranslated UI rather than using the `nn` variant?  I honestly don't know, but doesn't seem so likely from where I stand.  If it's really the case though, you should raise the issue to the C/gettext library not to select it then.

---

Until I have more info, I feel like the only thing to do here is accept a new `nb` translation. We'll have to rely on user submission here, because none of "us" can write `nb` -- but we'd happily accept such a translation.

**Comment by b4n** [comment]
> Probably would be good if "somebody" provided a preliminary `nb.po` file rather than rushing the whole translation at string freeze.

Yes, and it's unlikely a large amount of the translation would be invalidated anyway, so most of it would still be usable by then.

**Comment by elextr** [comment]
From what I can see there has been no update of actual translation for `nn.po` since 2012.  Although the cost of having both is not high, perhaps it would be better to have one unified `no.po` rather than both `nn.po` and `nb.po` since there doesn't seem much interest in maintaining translations, probably because as the OP said elsewhere in real life Norwegian techos code in English.

**Comment by o-l-a-v** [comment]
I assumed you'd like to do the same with `nb_NO` and `nn_NO` as you already do with `en_GB`, `zh_CN` and `zh_TW`.

`nb_NO` and `nn_NO` are quite similar, so it's probably OK to have `nn_NO` apply if system language/culture is anything `NO` and no `nb_NO` yet exist.

But I'd rather have it default to english when my closest language is ~17% complete ( https://www.geany.org/contribute/translation/statistics/ ). Maybe some threshold should be set for when to use translation?

About contributing with translations: Would be great with a non-technical way to do it, maybe some crowdsourcing in browser solution can be looked into?

**Comment by elextr** [comment]
> because conflict of the two first letters

correct, `en` is a special case since the reference is in (some variant of) English whereas the extra is needed for `zh` to distinguish them.

> nb_NO and nn_NO are quite similar, so it's probably OK to have nn_NO apply if system language/culture is anything NO and no nb_NO yet exist.

The choice is made by the gettext library, not by any Geany code, so we can't influence it AFAIK.  You could probably research how it decides, but I suspect it might be simply a straight comparison of the `LANG` environment with the PO file name.

> But I'd rather have it default to english when my closest language is ~17% complete

Totally understandable.

> Maybe some threshold should be set for when to use translation?

I don't think thats something the Geany team should decide, remembering all translations are contributed, people spend their own time doing it, it seems rude to tell someone "no, we won't use your work as it does not reach the arbitrary x% number".  If a mixed UI is difficult to use the `LANG` can always be set to `en`.

> maybe some crowdsourcing in browser solution can be looked into?

Ubuntu used one of those and just had to remove its ISOs because a translation from such a source contained inappropriate material.  At least the current system requires a pull request which puts the submitter's GitHub id on it so its not totally anonymous.

**Comment by o-l-a-v** [comment]
| Ubuntu used one of those and just had to remove its ISOs because a translation from such a source contained inappropriate material. At least the current system requires a pull request which puts the submitter's GitHub id on it so its not totally anonymous.

That's not a valid reason not to consider it IMO. That's like saying you shouldn't use git because some project got compromised on whatever hosted git platform they use.

One doesn't neccessarily need to know development and git to be useful for this project, translation is a great example. I think that by requiring people that can help you translate to know Linux and git you're loosing out on contributions, is what I'm saying.

**Comment by elextr** [comment]
There is no need to know Linux, editing PO files can just as easily be done on any platform, even with the github online editor which will happily make your PR for you.

Basically, the project is developed on github, at some point all changes must be made into pull requests and "somebody" has to do that.

**Comment by eht16** [comment]
> > Ubuntu used one of those and just had to remove its ISOs because a translation from such a source contained inappropriate material. At least the current system requires a pull request which puts the submitter's GitHub id on it so its not totally anonymous.
> 
> That's not a valid reason not to consider it IMO. That's like saying you shouldn't use git because some project got compromised on whatever hosted git platform they use.

Full ACK.

Recently I had a short chat with @frlan about using Transifex, Pootle, WebLate or so. Transifex won't make it because it's non-free. But the other platforms/solutions might improve the translation contributions. I think this does not necessarily have to exclude pull requests for manual review.

Anyway, this is off-topic here.

Based on @b4n's analysis, it seems the "nn.po" we use is correct and since we cannot change how gettext will match the translation to use to the system's locale, I think we cannot do anything here?

Choosing a translation based on its completeness sounds a bit weird to me.

**Comment by frlan** [comment]
To make it complete: by ISO 639 we have nb (Norsk Bokmål), nn (Norsk Nynorsk) and no (Norsk). I'm not an epxert in Norwegian language, but  I'm totally happy to update any language file that is sent to me -- this not even need a github account ;) 

I'm voting against removing any in favor of another as it's configurable with the typical LANG-environment variables on systems that run Geany (even for Windows somewhere in the Wiki I guess we had a short howto).

**Comment by frlan** [comment]
Oh as I don't see it in the thread: The _NO is only added if there are different versions as e.g. Portugues with pt and pt_BR or we have in en_US, en__GB, …
